### DISCUSSION TITLE
Decoding Failed in threejs after upgrading to gltf-transform@2.0.0

### DISCUSSION THREAD

**Discussion by oddnugget** [donmccurdy/glTF-Transform#488] [Q&A] [answered]
Since upgrading to v2 of glTF-Transform, whenever i try to open a gltf on https://gltf-viewer.donmccurdy.com/ that has been compressed with draco, i get the following error.

`THREE.DRACOLoader: Decoding failed: Failed to decode point attributes.`

The error only happens when I read a remote gltf file that has draco and merge it onto a newly created document with draco compression enabled. Interestingly it does not happen with Sequential Compression, only Edgebreaker, and neither does it happen 
when compressing an uncompressed local gltf.

**Comment by donmccurdy** [comment]
Hm, this may be difficult for me to reproduce without the assets, as compression is sensitive to some details in the geometry. Would you be able to share the two assets you're merging? Thanks!

**Reply by oddnugget** [reply]
Sure, thanks for the quick answer. 

So im merging this gltf with remote bins onto an empty document, 

<details>
<summary>a.gltf</summary>

```json
{
  "asset": { "version": "2.0", "generator": "THREE.GLTFExporter" },
  "scenes": [{ "name": "AuxScene", "nodes": [0] }],
  "scene": 0,
  "nodes": [
    {
      "name": "Sphere",
      "mesh": 0,
      "translation": [0, 0, 0],
      "rotation": [0, 0, 0, 1],
      "scale": [1, 1, 1]
    }
  ],
  "bufferViews": [{ "buffer": 0, "byteOffset": 0, "byteLength": 4518 }],
  "buffers": [
    {
      "name": "buffer",
      "byteLength": 4520,
      "uri": "https://development-cdn.sayduck.io/dvpt-or/shared_assets/files/000/000/002/dbfbfb9c559578743c32f717832d1df35cd176e061e225c380a0c55cc39e4345-original.bin"
    }
  ],
  "accessors": [
    {
      "componentType": 5123,
      "count": 2880,
      "type": "SCALAR",
      "min": [0],
      "max": [2879]
    },
    {
      "componentType": 5126,
      "count": 559,
      "max": [0.009999999776482582, 0.009999999776482582, 0.009999999776482582],
      "min": [
        -0.009999999776482582, -0.009999999776482582, -0.009999999776482582
      ],
      "type": "VEC3"
    },
    {
      "componentType": 5126,
      "count": 559,
      "max": [1, 1, 1],
      "min": [-1, -1, -1],
      "type": "VEC3"
    },
    {
      "componentType": 5126,
      "count": 559,
      "max": [1, 1],
      "min": [0, 0],
      "type": "VEC2"
    },
    {
      "componentType": 5126,
      "count": 559,
      "max": [1, 1],
      "min": [0, 0],
      "type": "VEC2"
    }
  ],
  "materials": [
    {
      "pbrMetallicRoughness": {
        "metallicFactor": 0,
        "roughnessFactor": 1,
        "baseColorFactor": [1, 1, 1, 1]
      },
      "emissiveFactor": [0, 0, 0],
      "alphaMode": "OPAQUE",
      "doubleSided": false
    }
  ],
  "meshes": [
    {
      "primitives": [
        {
          "mode": 4,
          "attributes": {
            "POSITION": 1,
            "NORMAL": 2,
            "TEXCOORD_0": 3,
            "TEXCOORD_1": 4
          },
          "material": 0,
          "indices": 0,
          "extensions": {
            "KHR_draco_mesh_compression": {
              "bufferView": 0,
              "attributes": {
                "POSITION": 0,
                "NORMAL": 1,
                "TEXCOORD_0": 2,
                "TEXCOORD_1": 3
              }
            }
          }
        }
      ]
    }
  ],
  "extensionsRequired": ["KHR_draco_mesh_compression"],
  "extensionsUsed": ["KHR_draco_mesh_compression"]
}
```

</details>

And then im merging these two local gltfs that only contain materials onto that as well


<details>
<summary>b.gltf</summary>

```json
{
  "asset": {
    "version": "2.0",
    "generator": "Sayduck API",
    "copyright": "2018 (c) Sayduck Oy"
  },
  "materials": [
    {
      "name": "Sphere 1 Material",
      "alphaMode": "OPAQUE",
      "alphaCutoff": 0.5,
      "emissiveFactor": [0.0, 0.0, 0.0],
      "pbrMetallicRoughness": {
        "baseColorFactor": [1.0, 0.0, 0.0, 1.0],
        "metallicFactor": 0.5,
        "roughnessFactor": 0.5
      },
      "extras": {
        "sayduck": { "uuid": "69202b30-6a24-013a-3eb8-07fb652ed1de" }
      }
    }
  ]
}
```
</details>


<details>
<summary>c.gltf</summary>

```json
{
  "asset": {
    "version": "2.0",
    "generator": "Sayduck API",
    "copyright": "2018 (c) Sayduck Oy"
  },
  "materials": [
    {
      "name": "Sphere 2 Material",
      "alphaMode": "OPAQUE",
      "alphaCutoff": 0.5,
      "emissiveFactor": [0.0, 0.0, 0.0],
      "pbrMetallicRoughness": {
        "baseColorFactor": [0.0, 0.30980392156862746, 1.0, 1.0],
        "metallicFactor": 0.5,
        "roughnessFactor": 0.5
      },
      "extras": {
        "sayduck": { "uuid": "4315fc70-6a27-013a-e768-19a3c4ad2b28" }
      }
    }
  ]
}

```
</details>

**Reply by donmccurdy** [reply]
Trying to reconstruct what might be happening here:

```
gltf-transform merge a.gltf b.gltf c.gltf merged.gltf --allow-http
gltf-transform draco merged.gltf merged_draco.gltf --allow-http
> error: [KHR_draco_mesh_compression] Decoding failure.
```

After https://github.com/donmccurdy/glTF-Transform/pull/477 and https://github.com/donmccurdy/glTF-Transform/issues/474, NodeIO skips writing any resources with a remote URL. I'd been assuming the use case was textures there — since textures probably won't change during processing — but in this case it's skipping writing the `.bin` resource, and that's not really possible, the `.bin` almost certainly will change.

Depending on how you're doing your merging, the way to fix this might be different. In the CLI, probably the remote `.bin` URL should just be replaced automatically with a local `.bin` path. For scripting, I think this case might need to just throw an error and let the user fix the URL.

> Interestingly it does not happen with Sequential Compression, only Edgebreaker...

In my test case above it'll happen for both methods, it seems to depend on the location of the `.bin` rather than the compression type. After the merge step, even a `copy` command (with no compression) will fail with the same error.

> EDIT: For now I've pushed https://github.com/donmccurdy/glTF-Transform/commit/e0fd00132b5b43bfbf2bdb1ec9cfaf5e3642d3eb, which should give a clearer error rather than writing a corrupted file.


[merge-test.zip](https://github.com/donmccurdy/glTF-Transform/files/8045556/merge-test.zip)

**Reply by oddnugget** [reply]
So it was basically failing because it can't write to a remote url, which makes sense. In my case the bins are allready compressed, and all I'm doing is merging materials and bins, and creating nodes, so i wouldn't actually have to run the result through draco. 

I guess it would still need to add the `extensionsUsed` property to the resulting gltf if the remote bins use draco though. Is that something that would have to be done manually?

**Reply by donmccurdy** [reply, marked as answer]
glTF Transform won't be able to write a file without also rewriting the associated `.bin` – there's too much to keep track of to be sure that the `.bin` hasn't changed, even if you're only merging materials. Moreover, loading the model in glTF Transform decompresses mesh data for editing. If you do something like `gltf-transform copy in_compressed.glb out.glb` you'll see this warning:

> Decoded KHR_draco_mesh_compression. Further compression will be lossy.

To merge two files _without_ decompressing anything I think you'd need to work directly with the raw JSON, and do pretty careful binary operations on the `.bin` files.  In general it is much easier to do processing before compression, and save compression until the last stage of the pipeline.

**Reply by oddnugget** [reply]
Allright. Looks like i will have to rethink my approach. Thanks for the help!
